File: ME5013/Fall2014/Exercises/InClass/Exercise13/laplace.py
#!/usr/bin/env python

import numpy
import Gnuplot
import matplotlib.pyplot as plt
import matplotlib.cm as cm
from mpl_toolkits.mplot3d import Axes3D
import sys

#Computes one step of the iteration
def computeTimeStep(u,dx,dy):
    dx2, dy2 = dx**2, dy**2
    dnr_inv = 0.5/(dx2 + dy2)
    u_old=u.copy()
    # The actual iteration
    u[1:-1, 1:-1] = ((u[0:-2, 1:-1] + u[2:, 1:-1])*dy2 +
                    (u[1:-1,0:-2] + u[1:-1, 2:])*dx2)*dnr_inv
    v = (u - u_old).flat
    return u,numpy.sqrt(numpy.dot(v,v))

num_points=100
dx=1.0/(num_points-1)
j=numpy.complex(0,1)
max_iter=1000000
num_iter=0
err_tol=10.0**(-6)


print("num_points: %d"%num_points)
print("dx: %f"%dx)
m=numpy.zeros((num_points,num_points),dtype=float)
pi_c=numpy.pi
x=numpy.r_[0.0:pi_c:num_points*j]
m[0,:]=numpy.sin(x)
m[num_points-1,:]=numpy.sin(x)

u = m

for i in range(max_iter):
        u,err=computeTimeStep(u,dx,dx)
        print "Iteration: "+str(i)
        print "Error: "+str(err)
        if err<err_tol:
                print "Converged"
                break

# plot using matplotlibb

x = y = numpy.linspace( 0.0, num_points*dx, num_points)
x, y = numpy.meshgrid(x, y)

fig = plt.figure()
ax = fig.gca(projection='3d')
surf = ax.plot_surface(x,y,u, rstride=1, cstride=1, cmap=cm, linewidth=0, antialiased=False)
ax.set_zlim(-0.01,1.01)
plt.show()

File: ME5013/Fall2014/Exercises/InClass/Exercise15/Part1/laplace.c
#include <math.h>
#include <stdio.h>
#include "laplace.h"

double timestep(double *u, int nx, int ny, double dx, double dy)
{
    double tmp, err, diff, dx2, dy2, dnr_inv;
    dx2=dx*dx;
    dy2=dy*dy;
    dnr_inv=0.5/(dx2+dy2);
    err=0.0;
    int i,j;

    for (i=1; i<nx-1; i++){
        for (j=1; j<ny-1; ++j){
            tmp = u[i*nx+j];
            u[i*nx+j] = ((u[(i-1)*nx+j] + u[(i+1)*nx+j])*dy2 + 
                    (u[i*nx+j-1] + u[i*nx+j+1])*dx2)*dnr_inv;
            diff = u[i*nx+j] - tmp;
            err += diff*diff;
        }
    }

    return sqrt(err);
}

double solve_in_C(double *u, int nx, int ny, double dx, double dy, int max_iter)
{
    double err;
    int iter;
    err = 1;
        for (iter=0; iter < max_iter; iter++){
            if(err < 1e-6)
                break;
            err = timestep(u,nx,ny,dx,dy);
        }

    return err;   
}
